service/iot-service-client/src/main/java/com/microsoft/azure/sdk/iot/service/transport/amqps/AmqpFeedbackReceivedHandler.java

```diff
@@ -154,49 +193,34 @@ public void onDelivery(Event event)
-
-        // Codes_SRS_SERVICE_SDK_JAVA_AMQPFEEDBACKRECEIVEDHANDLER_12_013: [The event handler shall create a Receiver (Proton) object and set the protocol tag on it to a predefined constant]
-        // Codes_SRS_SERVICE_SDK_JAVA_AMQPFEEDBACKRECEIVEDHANDLER_15_017: [The Receiver object shall have the properties set to service client version identifier.]
-        Map<Symbol, Object> properties = new HashMap<>();
-        properties.put(Symbol.getSymbol(TransportUtils.versionIdentifierKey), TransportUtils.USER_AGENT_STRING);
-        feedbackReceiverLink = ssn.receiver(RECEIVE_TAG);
-        feedbackReceiverLink.setProperties(properties);
-
-        // Codes_SRS_SERVICE_SDK_JAVA_AMQPFEEDBACKRECEIVEDHANDLER_12_014: [The event handler shall open the Connection, the Session and the Receiver object]
-        log.debug("Opening connection, session and link for amqp feedback receiver");
-        conn.open();
-        ssn.open();
-        feedbackReceiverLink.open();
-    }
-
-    @Override
-    public void onLinkInit(Event event)
+    public void onAuthenticationSucceeded()
     {
-        // Codes_SRS_SERVICE_SDK_JAVA_AMQPFEEDBACKRECEIVEDHANDLER_12_015: [The event handler shall create a new Target (Proton) object using the given endpoint address]
-        // Codes_SRS_SERVICE_SDK_JAVA_AMQPFEEDBACKRECEIVEDHANDLER_12_016: [The event handler shall get the Link (Proton) object and set its target to the created Target (Proton) object]
-        Link link = event.getLink();
-        if (event.getLink().getName().equals(RECEIVE_TAG))
+        // Only open the session and receiver link if this authentication was for the first open. This callback
+        // will be executed again after every proactive renewal, but nothing needs to be done after a proactive renewal
+        if (feedbackReceiverLink == null)
         {
+            // Every session or link could have their own handler(s) if we
+            // wanted simply by adding the handler to the given session
+            // or link
+
+            Session ssn = this.connection.session();
+
+            // If a link doesn't have an event handler, the events go to
+            // its parent session. If the session doesn't have a handler
+            // the events go to its parent connection. If the connection
+            // doesn't have a handler, the events go to the reactor.
+
+            Map<Symbol, Object> properties = new HashMap<>();
+            properties.put(Symbol.getSymbol(TransportUtils.versionIdentifierKey), TransportUtils.USER_AGENT_STRING);
+            feedbackReceiverLink = ssn.receiver(RECEIVE_TAG);
+            feedbackReceiverLink.setProperties(properties);
+
+            log.debug("Opening connection, session and link for amqp feedback receiver");
+            ssn.open();
+            feedbackReceiverLink.open();
             Source source = new Source();
             source.setAddress(ENDPOINT);
-            link.setSource(source);
+            feedbackReceiverLink.setSource(source);
```

Comment:
If I have understood this correctly, this.connection.session() => open session and  feedbackReceiverLink.open(); => open link ?

Reply:
Yep. Once CBS auth has completed successfully, we respond by creating a session and a receiver link in that session, and opening both. That session/link combo are responsible for listening for feedback messages from the service afterwards.

Reply:
You'll see similar code in the c2d handler and file upload notification handler, but with some slightly different link addresses